modules/utxo-lib/src/bitgo/wallet/Psbt.ts

```diff
@@ -134,3 +158,148 @@ export function signWalletPsbt(
     psbt.signInputHD(inputIndex, signer);
   }
 }
+
+function classifyScriptType(input: PsbtInput): ScriptType2Of3 | undefined {
+  let scriptType: ScriptType2Of3 | undefined;
+  if (Buffer.isBuffer(input.redeemScript) && Buffer.isBuffer(input.witnessScript)) {
+    scriptType = 'p2shP2wsh';
+  } else if (Buffer.isBuffer(input.redeemScript)) {
+    scriptType = 'p2sh';
+  } else if (Buffer.isBuffer(input.witnessScript)) {
+    scriptType = 'p2wsh';
+  }
+  if (Array.isArray(input.tapLeafScript) && input.tapLeafScript.length > 0) {
+    if (scriptType) {
+      throw new Error(`Found both ${scriptType} and p2tr PSBT metadata.`); // can be tested
+    }
+    if (input.tapLeafScript.length > 1) {
+      throw new Error(`Bitgo only supports a single tap leaf script per input.`); // can be tested
+    }
+    scriptType = 'p2tr';
+  }
+  return scriptType;
+}
+
+function parseSignatures(input: PsbtInput, scriptType: ScriptType2Of3): Signatures {
+  const validate = (sig: Buffer): Buffer => {
+    if (Buffer.isBuffer(sig)) {
+      return sig;
+    }
+    throw new Error('Invalid signature type');
+  };
+
+  let signatures: [Buffer] | [Buffer, Buffer];
+  if (scriptType === 'p2tr') {
+    if (input.partialSig && input.partialSig.length > 0) {
+      throw new Error('Invalid PSBT signature state');
+    }
+    if (!input.tapScriptSig || input.tapScriptSig.length === 0) {
+      return undefined;
+    }
+    if (input.tapScriptSig.length > 2) {
+      throw new Error('unexpected signature count'); // can be tested?
+    }
+    if (input.tapScriptSig.length === 1) {
+      signatures = [validate(input.tapScriptSig[0].signature)];
+    } else {
+      signatures = [validate(input.tapScriptSig[0].signature), validate(input.tapScriptSig[1].signature)];
+    }
+  } else {
+    if (input.tapScriptSig && input.tapScriptSig.length > 0) {
+      throw new Error('Invalid PSBT signature state');
+    }
+    if (!input.partialSig || input.partialSig.length === 0) {
+      return undefined;
+    }
+    if (input.partialSig.length > 2) {
+      throw new Error('unexpected signature count'); // can be tested?
+    }
+    if (input.partialSig.length === 1) {
+      signatures = [validate(input.partialSig[0].signature)];
+    } else {
+      signatures = [validate(input.partialSig[0].signature), validate(input.partialSig[1].signature)];
+    }
+  }
+  return signatures;
+}
+
+function parseScript(input: PsbtInput, scriptType: ScriptType2Of3) {
```

Comment:
even though it is internal, a return type helps understanding

Reply:
Done